<a href="https://colab.research.google.com/github/sowmyasoundar2610/data-analytics/blob/main/Air_pollution_Time_Series_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Topics Practiced**


1.Treating missing values

2.Treating garbage values

3.Date formatting- - pd.to_datetime()

**About the data**

The dataset contains 9358 instances of hourly averaged responses from an array of 5 metal oxide chemical sensors embedded in an Air Quality Chemical Multisensor Device.

The device was located on the field in a significantly polluted area, at road level, within an Italian city. Data were recorded from March 2004 to February 2005 

In [1]:
#Importing the required modules and loading the time-series dataset on air quality. 
import numpy as np
import pandas as pd
import datetime

csv_file = 'https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/air-quality/AirQualityUCI.csv'
df = pd.read_csv(csv_file, sep=';')
df.head()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00    2,6       1360.0     150.0     11,9         1046.0   
1  10/03/2004  19.00.00      2       1292.0     112.0      9,4          955.0   
2  10/03/2004  20.00.00    2,2       1402.0      88.0      9,0          939.0   
3  10/03/2004  21.00.00    2,2       1376.0      80.0      9,2          948.0   
4  10/03/2004  22.00.00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Unnamed: 15  Unnamed: 16  
0  0,7578          NaN          NaN  
1  0,7255          NaN          NaN  
2  0,7502          NaN          NaN  
3  0,7867          NaN          NaN  
4  0,7888          NaN          NaN

In [2]:
# Applying the 'info()' function on the 'df'
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   object 
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   object 
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   object 
 13  RH             9357 non-null   object 
 14  AH             9357 non-null   object 
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(10), object(7)
memory usage: 1.2+ MB


In [3]:
# Checking for the missing values in the 'df'.
df.isnull().sum()

Date              114
Time              114
CO(GT)            114
PT08.S1(CO)       114
NMHC(GT)          114
C6H6(GT)          114
PT08.S2(NMHC)     114
NOx(GT)           114
PT08.S3(NOx)      114
NO2(GT)           114
PT08.S4(NO2)      114
PT08.S5(O3)       114
T                 114
RH                114
AH                114
Unnamed: 15      9471
Unnamed: 16      9471
dtype: int64

All the values contained in the `Unnamed: 15 and Unnamed: 16` columns are the missing (or null) values. The other columns contain 114 null values. So, let's drop the last two columns using the `drop()` function.

In [4]:
#Dropping the 'Unnamed: 15 and Unnamed: 16' columns from the 'df'.
df = df.drop(columns=['Unnamed: 15', 'Unnamed: 16'], axis=1) 

In [5]:
# Printing the rows missing in the 'Date' column.
df[df['Date'].isnull() == True]

Date Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  NOx(GT)  \
9357  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9358  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9359  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9360  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9361  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
...   ...  ...    ...          ...       ...      ...            ...      ...   
9466  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9467  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9468  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9469  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   
9470  NaN  NaN    NaN          NaN       NaN      NaN            NaN      NaN   

      PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)    T   RH   AH  
9357           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9358           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9359           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9360           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9361           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
...            ...      ...           ...          ...  ...  ...  ...  
9466           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9467           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9468           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9469           NaN      NaN           NaN          NaN  NaN  NaN  NaN  
9470           NaN      NaN           NaN          NaN  NaN  NaN  NaN  

[114 rows x 15 columns]

In [6]:
#Drop the rows containing at least one null value in the 'df' DataFrame using the 'dropna()' function.
df = df.dropna()

In [7]:
#Checking whether all the missing values are removed or not.
df.isnull().sum()

Date             0
Time             0
CO(GT)           0
PT08.S1(CO)      0
NMHC(GT)         0
C6H6(GT)         0
PT08.S2(NMHC)    0
NOx(GT)          0
PT08.S3(NOx)     0
NO2(GT)          0
PT08.S4(NO2)     0
PT08.S5(O3)      0
T                0
RH               0
AH               0
dtype: int64

**Parsing datetime Values**


Going to convert the values stored in the Date and Time columns to datetime values so that later I can sort the DataFrame in the chronological order.

In [8]:
# Concatenating the values stored in the 'Date' and 'Time' columns and store them into a new Pandas series.
dt_series = pd.Series(data = [item.split("/")[2] + "-" + item.split("/")[1] + "-" + item.split("/")[0] for item in df['Date']], index=df.index) + ' ' + pd.Series(data=[str(item).replace(".", ":") for item in df['Time']], index=df.index)
dt_series = pd.to_datetime(dt_series)
dt_series

0      2004-03-10 18:00:00
1      2004-03-10 19:00:00
2      2004-03-10 20:00:00
3      2004-03-10 21:00:00
4      2004-03-10 22:00:00
               ...        
9352   2005-04-04 10:00:00
9353   2005-04-04 11:00:00
9354   2005-04-04 12:00:00
9355   2005-04-04 13:00:00
9356   2005-04-04 14:00:00
Length: 9357, dtype: datetime64[ns]

In [9]:
#Removing the Date & Time columns from the DataFrame and insert the 'dt_series' in it.
df = df.drop(columns=['Date', 'Time'], axis=1)
df.insert(loc=0, column='DateTime', value=dt_series)

Adding four more columns to the DataFrame. They will contain the year, month, day and day-name values for each observation on the air pollutants, temperature, relative humidity and absolute humidity.

In [10]:
#Pandas series containing the year values as integers.
year_series = dt_series.dt.year
year_series # year as integer

0       2004
1       2004
2       2004
3       2004
4       2004
        ... 
9352    2005
9353    2005
9354    2005
9355    2005
9356    2005
Length: 9357, dtype: int64

In [11]:
#Pandas series containing the month values as integers.
month_series = dt_series.dt.month
month_series # month as integer

0       3
1       3
2       3
3       3
4       3
       ..
9352    4
9353    4
9354    4
9355    4
9356    4
Length: 9357, dtype: int64

In [12]:
#Pandas series containing the day values as integers.
day_series = dt_series.dt.day
day_series # day as integer

0       10
1       10
2       10
3       10
4       10
        ..
9352     4
9353     4
9354     4
9355     4
9356     4
Length: 9357, dtype: int64

In [13]:
#Pandas series containing the days of a week, i.e., Monday, Tuesday, Wednesday etc.
day_name_series = dt_series.dt.day_name()
day_name_series

0       Wednesday
1       Wednesday
2       Wednesday
3       Wednesday
4       Wednesday
          ...    
9352       Monday
9353       Monday
9354       Monday
9355       Monday
9356       Monday
Length: 9357, dtype: object

In [14]:
#Adding the 'Year', 'Month', 'Day' and 'Day Name' columns to the DataFrame.
df['Year'] = year_series
df['Month'] = month_series
df['Day'] = day_series
df['Day Name'] = day_name_series

In [15]:
# Displaying the first five rows of the DataFrame after adding the new columns.
df.head()

DateTime CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0 2004-03-10 18:00:00    2,6       1360.0     150.0     11,9         1046.0   
1 2004-03-10 19:00:00      2       1292.0     112.0      9,4          955.0   
2 2004-03-10 20:00:00    2,2       1402.0      88.0      9,0          939.0   
3 2004-03-10 21:00:00    2,2       1376.0      80.0      9,2          948.0   
4 2004-03-10 22:00:00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Year  Month  Day   Day Name  
0  0,7578  2004      3   10  Wednesday  
1  0,7255  2004      3   10  Wednesday  
2  0,7502  2004      3   10  Wednesday  
3  0,7867  2004      3   10  Wednesday  
4  0,7888  2004      3   10  Wednesday

In [16]:
#Sorting the DataFrame by the 'DateTime' values in the ascending order
df = df.sort_values(by='DateTime')
df.head(10)

DateTime CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0 2004-03-10 18:00:00    2,6       1360.0     150.0     11,9         1046.0   
1 2004-03-10 19:00:00      2       1292.0     112.0      9,4          955.0   
2 2004-03-10 20:00:00    2,2       1402.0      88.0      9,0          939.0   
3 2004-03-10 21:00:00    2,2       1376.0      80.0      9,2          948.0   
4 2004-03-10 22:00:00    1,6       1272.0      51.0      6,5          836.0   
5 2004-03-10 23:00:00    1,2       1197.0      38.0      4,7          750.0   
6 2004-03-11 00:00:00    1,2       1185.0      31.0      3,6          690.0   
7 2004-03-11 01:00:00      1       1136.0      31.0      3,3          672.0   
8 2004-03-11 02:00:00    0,9       1094.0      24.0      2,3          609.0   
9 2004-03-11 03:00:00    0,6       1010.0      19.0      1,7          561.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   
5     89.0        1337.0     96.0        1393.0        949.0  11,2  59,2   
6     62.0        1462.0     77.0        1333.0        733.0  11,3  56,8   
7     62.0        1453.0     76.0        1333.0        730.0  10,7  60,0   
8     45.0        1579.0     60.0        1276.0        620.0  10,7  59,7   
9   -200.0        1705.0   -200.0        1235.0        501.0  10,3  60,2   

       AH  Year  Month  Day   Day Name  
0  0,7578  2004      3   10  Wednesday  
1  0,7255  2004      3   10  Wednesday  
2  0,7502  2004      3   10  Wednesday  
3  0,7867  2004      3   10  Wednesday  
4  0,7888  2004      3   10  Wednesday  
5  0,7848  2004      3   10  Wednesday  
6  0,7603  2004      3   11   Thursday  
7  0,7702  2004      3   11   Thursday  
8  0,7648  2004      3   11   Thursday  
9  0,7517  2004      3   11   Thursday

**Cleaning the data**


The  values in the `CO(GT), C6H6(GT), T,	RH` and	`AH` columns contain the commas in them. Going to replace the commas with periods (or dots). 

In [17]:
#Creating a function to replace the commas with periods in a Pandas series.
def comma_to_period(series):
    new_series = pd.Series(data=[float(str(item).replace(',', '.')) for item in series], index=df.index)
    return new_series

In [18]:
#Testing the 'comma_to_period()' function by correcting the values of 'CO(GT)' column but storing the output in a new variable.
new_series = comma_to_period(df['CO(GT)'])
new_series

0       2.6
1       2.0
2       2.2
3       2.2
4       1.6
       ... 
9352    3.1
9353    2.4
9354    2.4
9355    2.1
9356    2.2
Length: 9357, dtype: float64

In [19]:
#Applying the 'comma_to_period()' function on the ''CO(GT)', 'C6H6(GT)', 'T', 'RH' and 'AH' columns.
cols_to_correct = ['CO(GT)', 'C6H6(GT)', 'T', 'RH', 'AH'] # Create a list of column names.
for col in cols_to_correct: # Iterate through each column
    df[col] = comma_to_period(df[col]) # Replace the original column with the new series.

**Garbage value replacement**


The columns in the `df` DataFrame also contain `-200` value. It is a garbage value or just a random number to represent the further missing (or null) values in the DataFrame. Let me it with the most appropriate values for each column.

In [20]:
#How many rows contain -200 in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns?
neg_counts = [(col, df[df[col] == -200].shape[0]) for col in df.columns[1:-4]]
neg_counts

[('CO(GT)', 1683),
 ('PT08.S1(CO)', 366),
 ('NMHC(GT)', 8443),
 ('C6H6(GT)', 366),
 ('PT08.S2(NMHC)', 366),
 ('NOx(GT)', 1639),
 ('PT08.S3(NOx)', 366),
 ('NO2(GT)', 1642),
 ('PT08.S4(NO2)', 366),
 ('PT08.S5(O3)', 366),
 ('T', 366),
 ('RH', 366),
 ('AH', 366)]

In [21]:
#Finding out the percentage of rows containing -200 in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
neg_percent = [(col, round(df[df[col] == -200].shape[0] * 100 / df.shape[0], 2)) for col in df.columns[1:-4]]
neg_percent

[('CO(GT)', 17.99),
 ('PT08.S1(CO)', 3.91),
 ('NMHC(GT)', 90.23),
 ('C6H6(GT)', 3.91),
 ('PT08.S2(NMHC)', 3.91),
 ('NOx(GT)', 17.52),
 ('PT08.S3(NOx)', 3.91),
 ('NO2(GT)', 17.55),
 ('PT08.S4(NO2)', 3.91),
 ('PT08.S5(O3)', 3.91),
 ('T', 3.91),
 ('RH', 3.91),
 ('AH', 3.91)]

In [22]:
#Removing all the columns from the 'df' DataFrame containing more than 10% garbage value.
df = df.drop(columns=['NMHC(GT)', 'CO(GT)', 'NOx(GT)', 'NO2(GT)'], axis=1)
df.columns

Index(['DateTime', 'PT08.S1(CO)', 'C6H6(GT)', 'PT08.S2(NMHC)', 'PT08.S3(NOx)',
       'PT08.S4(NO2)', 'PT08.S5(O3)', 'T', 'RH', 'AH', 'Year', 'Month', 'Day',
       'Day Name'],
      dtype='object')

Let's replace the -200 value with the median values in all the columns except for the DateTime, Year, Month and Day columns.

Before that we should split the entire DataFrame in two different DataFrames because it contains data for two different years, i.e., 2004 and 2005. Then we should calculate the median values for each column for 2004 and 2005 separately. However, we can also first find out whether the median values are actually different for the two years. If they are not, then we don't need to split the DataFrame into two DataFrames for 2004 and 2005 data points.

In [ ]:
#Calculating the median values for the columns having indices between 1 and -4 (excluding -4) for the year 2004.
df.loc[df['Year'] == 2004, df.columns[1:-4]].describe()

In [ ]:
#Calculating the median values for the columns having indices between 1 and -4 (excluding -4) for the year 2005.
df.loc[df['Year'] == 2005, df.columns[1:-4]].describe()

:As we can see, the median values for 2004 and 2005 are different. Hence, we should split the `df` DataFrame into two different DataFrames. One for 2004 data points and another for 2005 data points.

In [ ]:
#Creating a new DataFrame containing records for the year 2004. Also, display the first five rows.
aq_2004_df = df[df['Year'] == 2004]
aq_2004_df.head()

In [ ]:
#Creating a new DataFrame containing records for the year 2005. Also, display the first five rows.
aq_2005_df = df[df['Year'] == 2005]
aq_2005_df.head()

In [ ]:
#Replace the -200 value with the median values for each column having indices between 1 and -4 (excluding -4) for the 2004 year DataFrame.
import warnings
warnings.filterwarnings('ignore')

for col in aq_2004_df.columns[1:-4]:
  median = aq_2004_df.loc[aq_2004_df[col] != -200, col].median() # Get the median value for each column after excluding -200.
  aq_2004_df[col] = aq_2004_df[col].replace(to_replace=-200, value=median)

In [ ]:
#Repeating the same for the 2005 year DataFrame.
for col in aq_2005_df.columns[1:-4]:
  median = aq_2005_df.loc[aq_2005_df[col] != -200, col].median()
  aq_2005_df[col] = aq_2005_df[col].replace(to_replace=-200, value=median)

In [ ]:
#Computing the number of rows containing '-200' in each column having indices between 1 and -4 (excluding -4) in the 2004 year DataFrame.
neg_val_2004 = [(col, aq_2004_df[aq_2004_df[col] == -200].shape[0]) for col in df.columns[1:-4]]
neg_val_2004

In [ ]:
#Again, calculating the percentage of rows containing '-200' in each column except for the 'DateTime', 'Year', 'Month' and 'Day' columns.
neg_val_2005 = [(col, aq_2005_df[aq_2005_df[col] == -200].shape[0]) for col in df.columns[1:-4]]
neg_val_2005